xds/googledirectpath/googlec2p.go

```diff
@@ -0,0 +1,175 @@
+/*
+ *
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package googledirectpath implements a resolver that configures xds to make
+// cloud to prod directpath connection.
+//
+// It's a combo of DNS and xDS resolvers. It delegates to DNS if
+// - not on GCE, or
+// - xDS bootstrap env var is set (so this client needs to do normal xDS, not
+// direct path, and clients with this scheme is not part of the xDS mesh).
+package googledirectpath
+
+import (
+	"fmt"
+	"time"
+
+	v3corepb "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/google"
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/internal/googlecloud"
+	internalgrpclog "google.golang.org/grpc/internal/grpclog"
+	"google.golang.org/grpc/resolver"
+	xdsclient "google.golang.org/grpc/xds/internal/client"
+	"google.golang.org/grpc/xds/internal/client/bootstrap"
+	"google.golang.org/grpc/xds/internal/env"
+	"google.golang.org/grpc/xds/internal/version"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/structpb"
+
+	_ "google.golang.org/grpc/xds" // To register xds resolvers and balancers.
+)
+
+const (
+	c2pScheme = "google-c2p"
+
+	tdURL          = "directpath-trafficdirector.googleapis.com"
+	httpReqTimeout = 10 * time.Second
+	zoneURL        = "http://metadata.google.internal/computeMetadata/v1/instance/zone"
+	ipv6URL        = "http://metadata.google.internal/computeMetadata/v1/instance/network-interfaces/0/ipv6s"
+
+	gRPCUserAgentName               = "gRPC Go"
+	clientFeatureNoOverprovisioning = "envoy.lb.does_not_support_overprovisioning"
+	ipv6CapableMetadataName         = "TRAFFICDIRECTOR_DIRECTPATH_C2P_IPV6_CAPABLE"
+
+	logPrefix = "[google-c2p-resolver]"
+
+	dnsName, xdsName = "dns", "xds"
+)
+
+type xdsClientInterface interface {
+	Close()
+}
+
+// For overriding in unittests.
+var (
+	onGCE = googlecloud.OnGCE
+
+	newClientWithConfig = func(config *bootstrap.Config) (xdsClientInterface, error) {
+		c, err := xdsclient.NewWithConfig(config)
+		return c, err
+	}
+
+	logger = internalgrpclog.NewPrefixLogger(grpclog.Component("directpath"), logPrefix)
+
+	defaultNode = &v3corepb.Node{
+		Id:                   "C2P",
+		Metadata:             nil, // To be set if ipv6 is enabled.
+		Locality:             nil, // To be set to the value from metadata.
```

Comment:
No need to initialize things to zero values, unless this is here for the comments (optionally remove explicit initializations and add a line that says `// Metadata and Locality are set later if ipv6 is enabled.`

Reply:
Done